### ISSUE TITLE
diff: only the first failing one is shown

### ISSUE THREAD

**Issue by avsm** [ocaml/dune#433]
In a rwo build with multiple files failing, only the first diff seems to be shown despite there being many promotable files:

```
<snip build log>
ocaml-topexpect examples/code/sexpr/sexp_opaque.topscript.stamp
ocaml-topexpect examples/code/variables-and-functions/main.topscript.stamp
ocaml-topexpect examples/code/variants/main.topscript.stamp
     patdiff (internal) (exit 1)
(cd _build/default && /Users/avsm/.opam/4.06.0/bin/patdiff -keep-whitespace -location-style omake -unrefined examples/code/classes/iter.topscript examples/code/classes/iter.topscript.corrected)
------ examples/code/classes/iter.topscript
++++++ examples/code/classes/iter.topscript.corrected
File "examples/code/classes/iter.topscript", line 68, characters 0-1:
 |
 |  method iterator : 'a iterator =
 |    new list_iterator v
 |end ;;
 |[%%expect ocaml {|
 |class ['a] stack :
 |  'a list ->
 |  object
 |    val mutable v : 'a list
 |    method iterator : 'a iterator
 |    method pop : 'a option
 |    method push : 'a -> unit
 |  end
 ||}];;
 |[@@@part "3"];;
 |let s = new stack [] ;;
-|[%%expect ocaml {|val s : '_a stack = <obj>|}];;
+|[%%expect ocaml {|val s : '_weak1 stack = <obj>|}];;
 |s#push 5 ;;
 |[%%expect ocaml {|- : unit = ()|}];;
 |s#push 4 ;;
 |[%%expect ocaml {|- : unit = ()|}];;
 |let it = s#iterator ;;
 |[%%expect ocaml {|val it : int iterator = <obj>|}];;
 |it#get ;;
 |[%%expect ocaml {|- : int = 4|}];;
 |it#next ;;
 |[%%expect ocaml {|- : unit = ()|}];;
 |it#get ;;
 |[%%expect ocaml {|- : int = 5|}];;
 |it#next ;;
 |[%%expect ocaml {|- : unit = ()|}];;
 |it#has_value ;;
 |[%%expect ocaml {|- : bool = false|}];;

Waiting for 3 jobs to finish.
make: *** [code] Error 1
$ jbuilder promote
Promoting _build/default/examples/code/async/main.topscript.corrected to examples/code/async/main.topscript.
Promoting _build/default/examples/code/classes/iter.topscript.corrected to examples/code/classes/iter.topscript.
Promoting _build/default/examples/code/command-line-parsing/command_types.topscript.corrected to examples/code/command-line-parsing/command_types.topscript.
Promoting _build/default/examples/code/command-line-parsing/step.topscript.corrected to examples/code/command-line-parsing/step.topscript.
Promoting _build/default/examples/code/guided-tour/main.topscript.corrected to examples/code/guided-tour/main.topscript.
Promoting _build/default/examples/code/imperative-programming/printf.topscript.corrected to examples/code/imperative-programming/printf.topscript.
Promoting _build/default/examples/code/imperative-programming/value_restriction.topscript.corrected to examples/code/imperative-programming/value_restriction.topscript.
Promoting _build/default/examples/code/imperative-programming/weak.topscript.corrected to examples/code/imperative-programming/weak.topscript.
Promoting _build/default/examples/code/maps-and-hash-tables/main.topscript.corrected to examples/code/maps-and-hash-tables/main.topscript.
Promoting _build/default/examples/code/records/main.topscript.corrected to examples/code/records/main.topscript.
Promoting _build/default/examples/code/records/main2.topscript.corrected to examples/code/records/main2.topscript.
```

Shouldn't all the diffs be shown?  Also, is it possible to have a command to show the diffs if you miss them in the build log? That is, something to run before `jbuilder promote` to manually inspect what is about to change before the mutation occurs, but without any further building.

**Comment by diml** [comment]
Currently jbuilder stops at the first error. We need to implement the equivalent of `make -k`.

> Shouldn't all the diffs be shown? Also, is it possible to have a command to show the diffs if you miss them in the build log? That is, something to run before jbuilder promote to manually inspect what is about to change before the mutation occurs, but without any further building.

That seems fine. I suppose we can have:

- `jbuilder promote --dry-run` to only show the list of files to copy
- `jbuilder promote --diffs` to only show the diffs again

**Comment by avsm** [comment]
A `make -k` mode would also make jbuilder slightly more deterministic to use in practise.  Right now when there is more than one failure, you effectively get a random one every run due to the parallelism.

**Comment by diml** [comment]
Indeed. It should probably the default in fact. It shouldn't be too hard to implemented, it's probably just adding a bunch of `Result.t` there and there. I'll try to have a look

**Comment by diml** [comment]
Done in #477

**Comment by diml** [comment]
Fixed by #477
